### ISSUE TITLE
Added compiler flags in conanfile.py but doesn't build with those flags? 

### ISSUE THREAD

**Issue by nimit25** [conan-io/conan#10891]
I have this in my conanfile.py
```
def package_info(self):
        self.cpp_info.cxxflags.append("-fstack-protector-strong")
```
However, I don't see these flags in the build. Is there an alternative way to add this? 
Should I just add the flags in CmakeLists.txt instead? 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Please read carefully the docs. The ``package_info()`` is information for consumers of the recipe, not for the same recipe. It is unusual that a recipe will define ``cxxflags`` for its consumers too, that might only happen for some recipes like packaged toolchains.

For defining flags for the self, current recipe, you can:

- Define them directly in your build system (note that Conan is not a build system) scripts, like CMakeLists.txt
- Pass them in the "toolchains", like the new ``CMakeToolchain``. This is the way to go if you want to parameterize, conditonally the injection of flags.
- You might be interested in the new global conf in https://github.com/conan-io/conan/pull/10800, which will be released in 1.47. The idea is that you could define the flag directly in your profiles and it will be injected by the toolchains (not only CMakeToolchain, but other toolchains too).

**Comment by franramirez688** [comment]
Conan 1.47 brings the chance to add it easily via `conf` so it's not needed to specify it in your conanfile.py.
Adding this configuration to your profile will be enough:
```
[conf]
tools.build:cxxflags=["-fstack-protector-strong"]
```
And it'll be automatically applied by the `conan_toolchain.cmake`.

**Comment by nimit25** [comment]
Is there some documentation for the above mentioned comment? I get a syntax error for ":" in the file

**Comment by memsharded** [comment]
Can you clarify what you are trying @nimit25?

Adding the above lines in your profile file and passing it with ``--profile=myprofile`` shouldn't error for ":". Alternatively, you can also try in the command line with ``-c tools.build:cxxflags=["xxxx"]``

**Comment by nimit25** [comment]
I added the above line in conanfile.py. I don't have a profile file in the workspace unfortunately

**Comment by memsharded** [comment]
> I added the above line in conanfile.py. I don't have a profile file in the workspace unfortunately

No, you cannot add that in the ``conanfile.py``. Just create a profile text file with the line, no need to be in the workspace, it can actually be in your Conan cache, you can put it in a repo and manage it with ``conan config install``. Or you can pass the config in command line. To summarize, the options:

- Adding flags in CMakeLists.txt. There is nothing wrong with this, this is probably recommended for many cases where the flag is always to be defined.
- When things are to be configured externally, configuration is defined in profiles, not in conanfile.py. All mechanisms, like settings, options, conf... are defined always in profiles (or command line).
   - External configuration can be customized in different ways. You can create custom settings, like "high-level" ``security: [enabled, disabled, extra, ..]`` and then use something like (not real, just example) ``if self.settings.security == "extra": cmaketoolchain.variables["INJECT_SEC"]="ON"`` (and then let the CMakeLists to parameterize adding flags depending on the cmake variable.
   - Inject directly flags via the new ``conf`` like ``tools.build:cxxflags``

**Comment by nimit25** [comment]
Is there an example package online I can look at for this or some sample documentation? A bit confused how to use this. I want to set it like 
os.setting.os == "Windows": 
      add these flags

**Comment by memsharded** [comment]
I don't know what you mean with ``os.setting.os == "Windows":`` That doesn't make sense, I guess you mean something like ``if self.settings.os == "Windows"``?

A conditional on the OS is not enough to require Conan configuration. You can very safely do something like:

```cmake
if(WIN32)
    # add flags here only for windows
endif()
```
Conan should be used to add different flags if you want to build, create and manage different binaries, some binaries with the flags enabled, and some binaries without the flags.

Even if you want to add flags, lets say only for Windows, that is typically managed with different profiles. You will have a ``windows`` profile, and a ``linux`` profile. And you will only add the respective flags in the respective profile and use it with ``conan ... -pr=windows``.

Finally, if you want to drive the behavior of some CMakeLists, based on some customization, like options, conf, or custom settings, the way is something like:

```python
def generate(self):
     tc = CMakeToolchain(self)
     if self.settings.security == "extra":
           tc.variables["MY_ADD_FLAG"] = "ON"
     tc.generate()
```
And your CMakeLists.txt will contain something like (pseudocode)
```cmake
if ("${MY_ADD_FLAG}")
      target_compile-options(mytarget ... -myflag1 -myflag2)
endif()
```

There are +1200 recipes, including very challenging to build open source libraries in https://github.com/conan-io/conan-center-index. But these are probably a bit complicated. For most cases you should be good starting with a ``conan new hello/0.1 --template=cmake_lib`` and doing small modifications while learning.

**Comment by rturrado** [comment]
> 

Hi @memsharded , I have just tried that syntax, `-c tools.build:cxxflags=["-D_LIBCPP_DISABLE_AVAILABILITY"]`, in a project of my own, with Conan 2.0.9, and I've got this error:

```
tools.build:cxxflags must be a list-like object. The value '[-D_LIBCPP_DISABLE_AVAILABILITY]' introduced is a str object
```

This is my whole line, in a `test.yml`:

```
conan build . -s:h compiler.cppstd=20 -s:h libqasm/*:build_type=Release -o libqasm/*:build_tests=True -o libqasm/*:compat=True -o libqasm/*:asan_enabled=True -b missing -c tools.build:cxxflags=["-D_LIBCPP_DISABLE_AVAILABILITY"]
```

**Comment by memsharded** [comment]
Hi @rturrado 

I think that you are experiencing shell expansion and parsing issues of your terminal, which doesn't understand Python lists.

You might need to do something like:
```
-c tools.build:cxxflags="['-D_LIBCPP_DISABLE_AVAILABILITY']"'
```
Note the external quotes to the list, then the internal single quotes for Python strings of every element of the list.

**Comment by rturrado** [comment]
Many thanks @memsharded , I ended up adding the flag from a `setup.py` file instead of from the `test.yml` above mentioned, and it is now working!

So yes, as you say, it has had to do with shell expansion, quotes, and so on.

Appreciated!

**Comment by memsharded** [comment]
Any further comment on https://github.com/conan-io/conan/issues/10891#issuecomment-1088067155 @nimit25 ?

Maybe the questions is addressed and we can close the ticket?

**Comment by memsharded** [comment]
Closing this ticket as outdated and staled, not followed up.

Please open new tickets for any further question or issue, thanks for the feedback!
